src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,639 @@
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance, but it is not considered thread-safe.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param derivationParameterSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters.
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationParameterSpec}
+     *     results in something invalid, ie - a key of inappropriate length
+     *     for the specified algorithm
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
+     * @throws NullPointerException
+     *     if {@code alg} or {@code derivationParameterSpec} is null
+     */
+    public SecretKey deriveKey(String alg,
+                               AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException, NoSuchAlgorithmException {
+        if (alg == null) {
+            throw new NullPointerException(
+                "the algorithm for the SecretKey return value should not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                "the algorithm for the SecretKey return value should not be "
+                + "empty");
+        }
+        Objects.requireNonNull(derivationParameterSpec);
+        if (spi != null) {
+            return spi.engineDeriveKey(alg, derivationParameterSpec);
+        } else {
+            return (SecretKey) chooseProvider(alg, derivationParameterSpec);
+        }
+    }
+
+    /**
+     * Obtains raw data from a key derivation function.
+     * <p>
+     * The {@code deriveData} method may be called multiple times on a
+     * particular {@code KDF} instance, but it is not considered thread-safe.
+     *
+     * @param derivationParameterSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return a byte array corresponding to the KDF output and according to
```

Comment:
I suggest `@return the derived key in its raw bytes`. This also implies my earlier suggestion on the relation between the output of the 2 derive methods.